core/controllers/access_validators_test.py

```diff
@@ -469,3 +471,86 @@ def test_validation_returns_false_if_given_user_is_non_existent(
             ), expected_status_int=404
         )
         self.logout()
+
+
+class TopicEditorPageAccessValidationHandlerTests(test_utils.GenericTestBase):
+    """Checks the access to the topic editor page and its rendering."""
+
+    def setUp(self) -> None:
+        super().setUp()
+        self.signup(self.CURRICULUM_ADMIN_EMAIL, self.CURRICULUM_ADMIN_USERNAME)
+        self.add_user_role(
+            self.CURRICULUM_ADMIN_USERNAME, feconf.ROLE_ID_CURRICULUM_ADMIN)
+
+        self.admin_id = self.get_user_id_from_email(self.CURRICULUM_ADMIN_EMAIL)
+
+        self.skill_id = skill_services.get_new_skill_id()
+        self.save_new_skill(
+            self.skill_id, self.admin_id, description='Skill Description')
+        self.skill_id_2 = skill_services.get_new_skill_id()
+        self.save_new_skill(
+            self.skill_id_2, self.admin_id, description='Skill Description 2')
+        self.topic_id = topic_fetchers.get_new_topic_id()
+        self.save_new_topic(
+            self.topic_id, self.admin_id, name='Name',
+            abbreviated_name='topic-one', url_fragment='topic-one',
+            description='Description', canonical_story_ids=[],
+            additional_story_ids=[],
+            uncategorized_skill_ids=[self.skill_id, self.skill_id_2],
+            subtopics=[], next_subtopic_id=1)
+
+    def test_access_topic_editor_page_without_logging_in(self) -> None:
+        self.get_json(
+            '%s/can_access_topic_editor/%s' % (
+                ACCESS_VALIDATION_HANDLER_PREFIX, self.topic_id
+            ), expected_status_int=401
+        )
+
+    def test_access_topic_editor_page_with_guest_user(self) -> None:
+        self.get_json(
+            '%s/can_access_topic_editor/%s' % (
+               ACCESS_VALIDATION_HANDLER_PREFIX, self.topic_id
+            ), expected_status_int=401
+        )
+
+        self.login(self.NEW_USER_EMAIL)
+        self.get_json(
+            '%s/can_access_topic_editor/%s' % (
+                ACCESS_VALIDATION_HANDLER_PREFIX, self.topic_id
+            ), expected_status_int=401
+        )
+        self.logout()
+
+    def test_access_topic_editor_page_with_curriculum_admin(
+        self
+    ) -> None:
+        self.login(self.CURRICULUM_ADMIN_EMAIL)
+        self.get_html_response(
+            '%s/can_access_topic_editor/%s' % (
+                ACCESS_VALIDATION_HANDLER_PREFIX, self.topic_id
+            ), expected_status_int=200
```

Comment:
Re: https://github.com/oppia/oppia/pull/19747#discussion_r1527218260 (Shouldn't you check what the response actually is, or assert that some expected string is contained in the response?) -- I suggest still checking something here. I looked at the code above and it seems like it doesn't follow indentation guidelines either so I am not sure if this was missed in previous reviews as well. Is there anything that would help identify the response as the right one (what does the response look like)?

> Also, Do we need to check, if we use the get_json response but the expected response is in HTML it gives an error and will be solved if we use get_html_response.

I don't really understand your question, sorry. What do you mean by "if we use the get_json response" and what is the actual scenario here?

Reply:
Sorry @seanlip, about my previous comment (It was not clear what I wanted to say )

What I understand so far is that for every request at `URL` there is a specific content type ( either `application/json`, `text/html` )

I check nearly all of the merged migration PR, and all of them directly use `get_html_response ` or `get_json`

Now here is what going on...

for `get_html_response` 
* checks for expected_errors https://github.com/oppia/oppia/blob/001e4c464289cefeac36445a7adfc05076316e07/core/tests/test_utils.py#L2806
https://github.com/oppia/oppia/blob/001e4c464289cefeac36445a7adfc05076316e07/core/tests/test_utils.py#L2832
* checks what the type of response we are getting from the URL ( either `application/json`, `text/html` )
https://github.com/oppia/oppia/blob/001e4c464289cefeac36445a7adfc05076316e07/core/tests/test_utils.py#L2834

which means if we use `get_json` instead of `get_html_response`  it will give an assertion error based on the above checks
* For the first check
```
    self.assertTrue(response.status_int >= 400)
AssertionError: False is not true
```
* for the second check
```
AssertionError: 'application/json' != 'text/html'
- application/json
+ text/html
```

Same for the `get_json`

https://github.com/oppia/oppia/blob/001e4c464289cefeac36445a7adfc05076316e07/core/tests/test_utils.py#L2982-L2984

https://github.com/oppia/oppia/blob/001e4c464289cefeac36445a7adfc05076316e07/core/tests/test_utils.py#L2936-L2938


So, we already have assertion checks in `get_json` and `get_html_response`based on what response we are getting from the URL.

Also, is it a good idea to first check the response type in access_handler_test for each request to URL and then use `get_json` or `get_html_response` ( knowing that both of these functions will already do the same job )?

Reply:
Hi @hrshkshri, I still don't understand what you are trying to say, sorry.

I get that we are doing generic checks for HTML and JSON responses. The first paragraph of my comment talks about checks on the **contents** of the response, not generic checks on the format. Does that help resolve some of the confusion here?

I also still can't understand what your last question means exactly -- it might be better to phrase your question as "should I go with option 1 or option 2" for this sort of thing, because I'm not sure what alternative you are asking me to help compare against.

Reply:
Hi @seanlip, Sorry for the late response...

> The first paragraph of my comment talks about checks on the contents of the response, not generic checks on the format. Does that help resolve some of the confusion here?

I understand the concern here, After checking all of the PRs ( Related to migration ), and the old `handler test` ( `...._test.py`[ /core/controllers ](https://github.com/oppia/oppia/tree/aa267c3e77b571cef9e6e7479119506f6651f55a/core/controllers )) which are related to the access validation route, what I am trying to say is that all of them follow the similar format ( At least what I understand )

Also, above I was trying to make a point that we shouldn't check if the response we are getting is either `application/json` or `text/html` since the `get_json()` and `get_html_response()` functions are doing that job.

But as you mentioned earlier (https://github.com/oppia/oppia/pull/19747#discussion_r1527218260) about checking on the contents of the response (not sure how to do that for now) could be the discussion here, also not to forget we are still migrating the pages, therefore, https://github.com/oppia/oppia/issues/19435 needs to be updated for the same (If we are planning to do that so).

WDYT?

Thanks

Reply:
Isn't it just checking `response.body`? There are some tests for that in the codebase. I think it's just adding a single line.

It's a reasonable question about whether we want to do that for all the tests. I'll defer to @DubeySandeep and @srijanreddy98 for that.

I still don't understand why the question of verifying json vs html came up, btw. Was it from my comment "what the response actually is"? If so, by that phrase, I meant the response content, not its type -- i.e. the same thing as I mentioned in the first paragraph of this response. Sorry for any confusion there.

Reply:
>I still don't understand why the question of verifying json vs html came up, btw. Was it from my comment "what the response actually is"? If so, by that phrase, I meant the response content, not its type -- i.e. the same thing as I mentioned in the first paragraph of this response. Sorry for any confusion there.

Sorry, I got a little confused by the comment on "response" and brought up the json vs html thing.

>It's a reasonable question about whether we want to do that for all the tests. I'll defer to @DubeySandeep and @srijanreddy98 for that.

/cc @DubeySandeep @srijanreddy98